model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientMapStorageProviderFactory.java

```diff
@@ -0,0 +1,223 @@
+                        }
+                        String password = config.get("password");
+                        if (password != null) {
+                            properties.put(AvailableSettings.JPA_JDBC_PASSWORD, password);
+                        }
+                    }
+
+                    String schema = config.get("schema");
+                    if (schema != null) {
+                        properties.put(JpaUtils.HIBERNATE_DEFAULT_SCHEMA, schema);
+                    }
+
+                    properties.put("hibernate.show_sql", config.getBoolean("showSql", false));
+                    properties.put("hibernate.format_sql", config.getBoolean("formatSql", true));
+                    properties.put("hibernate.dialect", config.get("driverDialect"));
+
+                    Connection connection = getConnection();
+                    try {
+                        printOperationalInfo(connection);
+
+                        customChanges(connection, schema, session, session.getProvider(MapJpaUpdaterProvider.class));
+
+                        logger.trace("Creating EntityManagerFactory");
+                        Collection<ClassLoader> classLoaders = new ArrayList<>();
+                        if (properties.containsKey(AvailableSettings.CLASSLOADERS)) {
+                            classLoaders.addAll((Collection<ClassLoader>) properties.get(AvailableSettings.CLASSLOADERS));
+                        }
+                        classLoaders.add(getClass().getClassLoader());
+                        properties.put(AvailableSettings.CLASSLOADERS, classLoaders);
+                        this.emf = JpaUtils.createEntityManagerFactory(session, unitName, properties, false);
+                        logger.trace("EntityManagerFactory created");
+
+                    } finally {
+                        // Close after creating EntityManagerFactory to prevent in-mem databases from closing
+                        if (connection != null) {
+                            try {
+                                connection.close();
+                            } catch (SQLException e) {
+                                logger.warn("Can't close connection", e);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private void customChanges(Connection connection, String schema, KeycloakSession session, MapJpaUpdaterProvider updater) {
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession lockSession) -> {
+            DBLockManager dbLockManager = new DBLockManager(lockSession);
+            DBLockProvider dbLock2 = dbLockManager.getDBLock();
+            dbLock2.waitForLock(DBLockProvider.Namespace.DATABASE);
+            try {
+                updater.update(ClientModel.class, connection, schema);
+            } finally {
+                dbLock2.releaseLock();
+            }
```

Comment:
TODO not in this PR: The locking needs to be handled by the provider factory on the area level, i.e. here for client tables. There should be no separate locking mechanism necessary, all the locking has to be done in here, and the locked area should not prevent locking and updating other area, e.g. realms or users.

Reply:
https://issues.redhat.com/browse/KEYCLOAK-19896